### PR TITLE
229 medical api first cut

### PR DESCRIPTION
**Pull request by Goncharo**
Ticket: metriport/metriport-internal#229

### Dependencies

- Downstream: https://github.com/metriport/metriport-internal/pull/364

### Description

_[Document your changes, give context for reviewers, add images/videos of UI changes]_

### Release Plan

- [ ] Merge this PR
- [ ] Merge downstream deps

### CODE REVIEW

File: api/app/src/sequelize/migrations/20230111000000-create-medical-models.ts
Comment (leite08) [inline api/app/src/sequelize/migrations/20230111000000-create-medical-models.ts:56]: That's weird, I just tried setting `uniqueKeys` on Sequelize `createTable`'s `options` param and it worked:

image

image

I'd actually refactor our `createTable` so it gets `options` (`transaction`, `uniqueKeys`) instead of just adding a new parameter - something like this:

```ts
export type CreateTableOptions = Omit<QueryInterfaceCreateTableOptions, "transaction"> &
  DeepNonNullable<Required<Pick<QueryInterfaceCreateTableOptions, "transaction">>>;

export const createTable = async (
  queryInterface: QueryInterface,
  tableName: string,
  tableDefinitions: ModelAttributes,
  options: CreateTableOptions
) => {
  await queryInterface.createTable(
    tableName,
    {
      ...tableDefinitions,
      ...defaultColumnsDef(),
    },
    options
  );
  await addUpdatedAtTrigger(queryInterface, options.transaction, tableName);
};
```

**Review by leite08 — COMMENTED** [review]
Main things to discuss:
- How to do FKs
- Whether to abstract the model structure on the DB or on the code as well

File: api/app/src/models/medical/facility.ts
Comment (Goncharo) [reply api/app/src/models/medical/facility.ts:29]: Agreed on both what is referenced + the naming, will update accordingly

File: api/app/src/models/medical/facility.ts
Comment (Goncharo) [reply api/app/src/models/medical/facility.ts:25]: We could - but then we'll have to query the id from the org every time the cx makes a facility request, to make sure they're the owners of the facility.

My 2 cents would be to leave the cxId in all models that belong to a cx.

File: api/app/src/models/medical/organization.ts
Comment (Goncharo) [reply api/app/src/models/medical/organization.ts:11]: Agreed (note the [TODOs about this](https://github.com/metriport/metriport/blob/3ba5b6bb1ac178b1ce891d8a7a5bffec818844fb/api/app/src/routes/medical/patient.ts#L27)) - to get this out the door quick you okay if I raise an issue to make this strongly typed after the initial customer onboarding?

File: api/app/src/routes/medical/facility.ts
Comment (Goncharo) [reply api/app/src/routes/medical/facility.ts:36]: I get the concern about multiple things being created at the same time, but if the concern is a race condition - we should be validating this with table locking behavior, and not retries/recursive functions.

Can you elaborate on what you mean here?

File: api/app/src/models/medical/organization.ts
Comment (Goncharo) [reply api/app/src/models/medical/organization.ts:40]: I get the concern about multiple things being created at the same time, but if the concern is a race condition - we should be validating this with table locking behavior, and not retries/recursive functions.

Can you elaborate on what you mean here?

File: api/app/src/routes/medical/facility.ts
Comment (Goncharo) [reply api/app/src/routes/medical/facility.ts:36]: The update never modifies the Org id - this is just used for the query filter.

Also, `getOrganizationOrFail()` expects a single org to be returned - multi-tenancy for a single cx is not supported

File: api/app/src/routes/medical/facility.ts
Comment (Goncharo) [reply api/app/src/routes/medical/facility.ts:68]: It will remove the need to add `facility.facilityId`, however we'll still want to do these conversions so that the data structure is flat and so we're not sending useless fields like `cxId`, `createdAt`, `updatedAt` etc.

File: api/app/src/models/medical/organization.ts
Comment (leite08) [reply api/app/src/models/medical/organization.ts:40]: I'm assuming there's no locking to generate this ID, since this is done in memory and I couldn't find anything that would indicate a semaphore/lock.

Retrying is, in theory, cheaper than locking the whole table or building some other type of lock to prevent more than one resource with the same ID.

I took a quick look at what Postgres have for this, [this](https://stackoverflow.com/a/13727025/2099911) seems like the best approach other than retrying.

File: api/app/src/routes/medical/patient.ts
Comment (Goncharo) [reply api/app/src/routes/medical/patient.ts:52]: As the "create or update" logic is <20 LOC, and most likely will only ever be used here, you okay to leave this until we actually use it in another part of the API?

File: api/app/src/sequelize/migrations/20230111000000-create-medical-models.ts
Comment (Goncharo) [reply api/app/src/sequelize/migrations/20230111000000-create-medical-models.ts:56]: Interesting, didn't know that there was another way to do this. What I was doing was something like:

```
      organizationId: {
        type: DataTypes.INTEGER,
        field: "organization_id",
        unique: "some_key"
      },
      facilityId: {
        type: DataTypes.INTEGER,
        field: "facility_id",
        unique: "some_key"
      },
```

Which according to Sequelize: "If true, the column will get a unique constraint. If a string is provided, the column will be part of a composite unique index. If multiple columns have the same string, they will be part of the same unique index" -> however no index was being created

Will update accordingly

File: api/app/src/models/medical/organization.ts
Comment (Goncharo) [reply api/app/src/models/medical/organization.ts:40]: Note that we can use `SERIAL` here -> but only b/c org IDs need to be unique across cxs. This approach won't work for facilities/patients, so will need to have another solution here

File: api/app/src/models/medical/facility.ts
Comment (leite08) [reply api/app/src/models/medical/facility.ts:25]: The problem with keeping the `cxId` on the model, IMO, is it pollutes the domain with unnecessary information...
- every entity/model would need this information
- we'd need to add/remove it as we move data across diff application layers

We could have a middleware checking ownership of data prior to the actual MW being executed.

Both have a cost:
- additional column
  - extra space on each table (storage)
  - extra data being moved from/to DB (performance)
  - DX wise, this requires massaging/work around for every operation we want to build on a given entity
- additional query
  - extra query for each request (performance)
  - DX wise, its something to be done once per entity.

IMO, the most expensive one for performance is the additional query, but I think its not a big deal; and the most expensive one for DX is additional column.

_DX = Developer eXperience_

File: api/app/src/models/medical/organization.ts
Comment (leite08) [reply api/app/src/models/medical/organization.ts:11]: Sure, I just thought we decided to have this abstraction on the model, not on the code. My concern with the code is the "we'll to it later" and it creeps as we build new things, bc we'll still want to be fast, and its likely that we'll see this as not adding value to the task/goal at hand.

But I'm cool with pushing as is for the onboarding of the first customer and create a ticket for this :)

File: api/app/src/command/medical/facility/update-facility.ts
Comment (leite08) [inline api/app/src/command/medical/facility/update-facility.ts:21]: It can equal to zero if no record was found, right?

File: api/app/src/routes/medical/facility.ts
Comment (leite08) [reply api/app/src/routes/medical/facility.ts:68]: `cxId` is part of [that larger convo](https://github.com/metriport/metriport/pull/111#discussion_r1112389025), but yeah, if it stays would need to be removed.

`createdAt` and `updatedAt` could be returned as well, no?

File: api/app/src/routes/medical/patient.ts
Comment (leite08) [reply api/app/src/routes/medical/patient.ts:52]: We can defs leave here, but its about defining patterns. Others might base off this, and it might not be small.

Yet, I'd argue its enough logic to warrant a command. :)

But cool if you want to push as is for the mvp

File: api/app/src/models/medical/facility.ts
Comment (Goncharo) [reply api/app/src/models/medical/facility.ts:25]: Additional things to consider for DX - I actually think removing the cxId will make the DX net worse than having it:

- making APIs for "get all patients for a customer" start looking a bit convoluted, as we'll always have to reference back to the org table to get the org ID for a cx - also there could be multiple orgs for a cx in the future in which this logic would have to be updated
- also, we're starting to go down the path of tribal/domain knowledge where a new dev has to understand the full domain picture to be able to deal with a single data type (ie what are all the cx's patients, or facilities)
- from an operations perspective, this also makes debugging + ops DML more convoluted - we will always have to again refer to the org table to filter for medical data belonging to a customer

A couple notes:

- the main dx point is needing to strip this from the response objects -> however we have DTOs elsewhere regardless and alternatively we could just return the cxId in the models as well if we want.
- when working on Aurora at AWS, the customer ID was always included in all the data models -> particularly for the Ops reason.

File: api/app/src/sequelize/migrations/20230111000000-create-medical-models.ts
Comment (leite08) [reply api/app/src/sequelize/migrations/20230111000000-create-medical-models.ts:56]: Yeah, found somewhere on SO people talking about these approaches, and the `unique` "label" not working. 👍

File: api/app/src/models/medical/facility.ts
Comment (Goncharo) [reply api/app/src/models/medical/facility.ts:25]: As discussed with @leite08 & @Orta21 :

"seems like the best way forward is what we discussed to not include the Org ID in the Facility/Patient model, and just include the customer ID - can't think of any disadvantages now aside from the multi-tenancy limitation (but we can always add this later).

So, we get the benefit of removing redundancy with extra data, and keep ability to reference a cx directly in the top-level models. Also no need to create a new customer table in OSS"

File: api/app/src/command/medical/facility/update-facility.ts
Comment (Goncharo) [reply api/app/src/command/medical/facility/update-facility.ts:21]: Good point - will update the error accordingly

File: api/app/src/models/medical/organization.ts
Comment (Goncharo) [reply api/app/src/models/medical/organization.ts:11]: Sounds good - note the later will come quickly as I'll add it to the scope of the issues that will need to implement the CW side of things

File: api/app/src/routes/medical/facility.ts
Comment (Goncharo) [reply api/app/src/routes/medical/facility.ts:68]: As per the convo, we're keeping cxId - also createdAt and updatedAt aren't needed atm so will leave them out

File: api/app/src/models/medical/organization.ts
Comment (Goncharo) [reply api/app/src/models/medical/organization.ts:40]: Will take care of this under https://github.com/metriport/metriport-internal/issues/366 before we ship the API to customers

**Review by Orta21 — APPROVED** [review]

File: api/app/src/routes/medical/facility.ts
Comment (leite08) [inline api/app/src/routes/medical/facility.ts:45]: Can we add a TODO to build a DTO here? It would be easy for someone to make a slight change on the return here and not change the `GET` below, for example, while both should be the same for consistency.

File: api/app/src/routes/medical/patient.ts
Comment (leite08) [inline api/app/src/routes/medical/patient.ts:73]: Ditto about adding a TODO to create a DTO

**Review by leite08 — APPROVED** [review]
Just a few questions and asks, the main think being the query by facilityIds and throwing 404 when nothing gets updated.

File: api/app/src/command/medical/facility/create-facility.ts
Comment (Goncharo) [reply api/app/src/command/medical/facility/create-facility.ts]: This is fixed - didn't work anyways

**Comment by Goncharo** [comment]
Will address remaining comments in next MAPI issue after merging to unblock parallel work

File: api/app/src/command/medical/patient/create-patient.ts
Comment (leite08) [reply api/app/src/command/medical/patient/create-patient.ts]: We don't have the Org ID on the Patient any more.

**Pull request merged**
